src/audio/clap/plugin_manager.cpp

```diff
@@ -0,0 +1,299 @@
+		}
+		return plugin_path;
+	}();
+
+#else // Windows, Linux
+
+	// We use the path of the dynamic-link library directly on Windows and
+	// Linux.
+	const auto library_path = plugin_path;
+#endif
+
+	const auto lib = dynlib_open(library_path);
+	if (!lib) {
+		LOG_ERR("CLAP: Error loading plugin library '%s'",
+		        library_path.string().c_str());
+		return {};
+	}
+
+	const auto plugin_entry = reinterpret_cast<const clap_plugin_entry_t*>(
+	        dynlib_get_symbol(lib, "clap_entry"));
+
+	if (!plugin_entry) {
+		dynlib_close(lib);
+		return {};
+	}
+
+	plugin_entry->init(reported_plugin_path.string().c_str());
+	return {lib, plugin_entry};
+}
+
+static std::vector<PluginInfo> get_plugin_infos(const std_fs::path& path,
+                                                const clap_plugin_entry_t* plugin_entry)
+{
+	const auto factory = static_cast<const clap_plugin_factory*>(
+	        plugin_entry->get_factory(CLAP_PLUGIN_FACTORY_ID));
+	assert(factory);
+
+	const auto num_plugins = factory->get_plugin_count(factory);
+	std::vector<PluginInfo> plugin_infos = {};
+
+	for (size_t plugin_index = 0; plugin_index < num_plugins; ++plugin_index) {
+		const auto desc = factory->get_plugin_descriptor(factory,
+		                                                 plugin_index);
+		if (desc) {
+			LOG_INFO("CLAP: Found plugin '%s'", desc->name);
+
+			const PluginInfo info = {path,
+			                         desc->id,
+			                         desc->name,
+			                         desc->description};
+
+			plugin_infos.emplace_back(info);
+		}
+	}
+
+	return plugin_infos;
+}
+
+void PluginManager::EnumeratePlugins()
+{
+	plugin_infos.clear();
+
+	constexpr auto OnlyRegularFiles = false;
+
+	for (const auto& [dir, plugin_names] :
+	     get_files_in_resource(PluginsDir, ".clap", OnlyRegularFiles)) {
+
+		LOG_DEBUG("CLAP: Enumerating CLAP plugins in '%s'",
+		          dir.string().c_str());
+
+		for (const auto& name : plugin_names) {
+			auto plugin_path = dir / name;
+
+			LOG_DEBUG("CLAP: Trying to load plugin library '%s'",
+			          plugin_path.string().c_str());
+
+			const auto [_, plugin_entry] = load_plugin_library(plugin_path);
+
+			if (!plugin_entry) {
+				LOG_WARNING("CLAP: Invalid plugin library '%s'",
+				            plugin_path.string().c_str());
+				continue;
+			}
+			const auto pi = get_plugin_infos(plugin_path, plugin_entry);
+
+			plugin_infos.insert(plugin_infos.end(), pi.begin(), pi.end());
+		}
+	}
+
+	plugins_enumerated = true;
+}
+
+std::vector<PluginInfo> PluginManager::GetPluginInfos()
+{
+	if (!plugins_enumerated) {
+		EnumeratePlugins();
+	}
+	return plugin_infos;
+}
+
+static bool validate_note_ports(const clap_plugin_t* plugin)
+{
+	const auto note_ports = static_cast<const clap_plugin_note_ports_t*>(
+	        plugin->get_extension(plugin, CLAP_EXT_NOTE_PORTS));
+
+	constexpr auto InputPort  = true;
+	constexpr auto OutputPort = false;
+
+	const auto num_in_ports = note_ports->count(plugin, InputPort);
```

Comment:
I'm playing with this CLAP API.  I've got a "Hello World" plugin I wrote in Rust that loads but I get a segfault here.  Should be a null check and return false if the extension is not supported.  The docs don't explicitly say that `get_extension` returns null on failure but I don't know what else a plugin would do if you request an extension that it doesn't have.

Reply:
Making the host more robust would be probably a job for later. I think the guiding principle should be making it work with actually useful plugins, not just any random hello world example 😆

If you're really interested in this, you should test your plugin with the [clap-validator](https://github.com/free-audio/clap-validator). The port extensions are kinda mandatory because without ports the plugin has no I/O, so it's rather useless. Having said all that, I can make a simple check & return here because it's easy but won't spend more time now on improving the robustness of the host.

Reply:
Well, I _did_ dive into it, after all 😎 The use of all port extensions are optional:

https://github.com/free-audio/clap/blob/main/include/clap/ext/note-ports.h#L8-L9

https://github.com/free-audio/clap/blob/main/include/clap/ext/audio-ports.h#L8-L10